crates/host/src/wasmbus/mod.rs

```diff
@@ -326,9 +343,79 @@ impl Invocation {
             msg,
             id,
             encoded_claims,
-            host_id,
+            host_id: host_id.to_string(),
         })
     }
+
+    /// A fully-qualified URL indicating the origin of the invocation
+    pub fn origin_url(&self) -> String {
+        self.origin.url()
+    }
+
+    /// A fully-qualified URL indicating the target of the invocation
+    pub fn target_url(&self) -> String {
+        format!("{}/{}", self.target.url(), self.operation)
+    }
+
+    /// The hash of the invocation's target, origin, and raw bytes
+    pub fn hash(&self) -> String {
+        invocation_hash(
+            self.target_url(),
+            self.origin_url(),
+            &self.operation,
+            &self.msg,
+        )
+    }
+
+    /// Validates the current invocation to ensure that the invocation claims have
+    /// not been forged, are not expired, etc
+    pub fn validate_antiforgery(&self, valid_issuers: &[String]) -> anyhow::Result<()> {
+        match KeyPair::from_public_key(&self.host_id) {
+            Ok(kp) if kp.key_pair_type() == KeyPairType::Server => (),
+            _ => bail!("invalid host ID on invocation: '{}'", self.host_id),
+        }
+
+        let vr = jwt::validate_token::<wascap::prelude::Invocation>(&self.encoded_claims)
+            .map_err(|e| anyhow!(e))?;
+        if vr.expired {
+            bail!("invocation claims token expired");
+        }
+        if !vr.signature_valid {
```

Comment:
Can we use `ensure!` for all of these please?

Reply:
for (en)sure, TIL about that macro